File: backend/services/vector_service/schemas/vector.py
"""Vector service Pydantic schemas."""
from typing import Optional

from pydantic import BaseModel, Field


class IndexRequest(BaseModel):
    paper_id: int = Field(..., gt=0)
    rebuild: bool = False


class IndexStatsData(BaseModel):
    paper_id: int
    engine: str
    vector_count: int
    dimension: int
    vectors_indexed: int
    index_path: str
    index_size_bytes: int
    bit_width: int
    status: str
    processing_time_ms: int


class SearchRequest(BaseModel):
    paper_id: int = Field(..., gt=0)
    query_embedding: list[float] = Field(..., min_length=1)
    top_k: int = Field(default=5, gt=0, le=100)
    allowlist: Optional[list[int]] = None


class SearchHit(BaseModel):
    chunk_id: int
    score: float
    paper_id: int
    chunk_index: int
    chunk_text: str


class SearchResultData(BaseModel):
    paper_id: int
    top_k: int
    engine: str
    results: list[SearchHit]
    search_latency_ms: int

File: frontend/src/api/vector.ts
import type { APIResponse } from "@/types";
import { apiFetch } from "./auth";

export interface IndexStats {
  paper_id: number;
  engine: string;
  vector_count: number;
  dimension: number;
  vectors_indexed: number;
  index_path: string;
  index_size_bytes: number;
  bit_width: number;
  status: string;
  processing_time_ms: number;
}

export interface SearchHit {
  chunk_id: number;
  score: number;
  paper_id: number;
  chunk_index: number;
  chunk_text: string;
}

export interface SearchResult {
  paper_id: number;
  top_k: number;
  engine: string;
  results: SearchHit[];
  search_latency_ms: number;
}

export const vectorApi = {
  buildIndex: (paperId: number, rebuild = false): Promise<APIResponse<IndexStats>> =>
    apiFetch<IndexStats>("/index", {
      method: "POST",
      body: JSON.stringify({ paper_id: paperId, rebuild }),
    }),

  search: (
    paperId: number,
    queryEmbedding: number[],
    topK = 5
  ): Promise<APIResponse<SearchResult>> =>
    apiFetch<SearchResult>("/search", {
      method: "POST",
      body: JSON.stringify({
        paper_id: paperId,
        query_embedding: queryEmbedding,
        top_k: topK,
      }),
    }),
};
